File: dynamic_listings/python/default/design_patterns/facade.py
class FirstService:
    # Implementation here...
    pass


class SecondService:
    # Implementation here...
    pass


class Facade:
    """
    self.class hides the complexities of using
    FirstService and SecondService from the user
    by "wrapping" them in a comfortable startAll
    function
    """
    service1: FirstService = None
    service2: SecondService = None

    def __init__(self) -> None:
        self.service1 = FirstService()
        self.service2 = FirstService()

    def startAll(self) -> bool:
        """
        The facade starts all the services and does
        some status checking, self.is hidden from the
        user.
        Returns true if all services started successfully
        false otherwise
        """
        firstServiceStarted = self.service1.start()
        if not firstServiceStarted:
            return False

        secondServiceStarted = self.service2.start()
        if not secondServiceStarted:
            return False

        # Here everything started successfully
        return True

File: dynamic_listings/rust/default/design_patterns/facade.rs
# #![allow(dead_code)]
struct FirstService {
    // Implementation here...
}
# impl FirstService {
#     fn new() -> Self { todo!() }
#     fn start(&self) -> bool { todo!() }
# }

struct SecondService {
    // Implementation here...
}
# impl SecondService {
#     fn new() -> Self { todo!() }
#     fn start(&self) -> bool { todo!() }
# }

struct Facade {
    /*
     * This class hides the complexities of using
     * FirstService and SecondService from the user
     * by "wrapping" them in a comfortable startAll
     * function
     */
    service1: FirstService,
    service2: SecondService,
}

impl Facade {
    pub fn new() -> Self {
        Self {
            service1: FirstService::new(),
            service2: SecondService::new(),
        }
    }

    fn start_all(&self) -> bool {
        /*
         * The facade starts all the services and does
         * some status checking, this is hidden from the
         * user.
         * Returns true if all services started successfully
         * false otherwise
         */
        let first_service_started = self.service1.start();
        if !first_service_started {
            return false;
        }
        let second_service_started = self.service2.start();
        if !second_service_started {
            return false;
        }
        // Here everything started successfully
        true
    }
}
